### PR TITLE
feat: add delay to processing of new Bitcoin blocks

### PR DESCRIPTION
**Pull request by aldur**
## Description

Closes: #701 

## Changes

- As suggested in #701, the coordinator now waits 15 seconds before processing new Bitcoin blocks.
- In tests, it doesn't wait (unless explicitly asked to).

## Testing Information

Run: `cargo test --lib --package signer should_wait_before_processing_bitcoin_blocks`.

## Checklist:

- [X] I have performed a self-review of my code
- [X] My changes generate no new warnings
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

Note that the following tests fail locally and on current `main` as well:

```txt
test network::libp2p::network::tests::swarm_rejects_connections_from_unknown_peers ... FAILED
test network::libp2p::network::tests::two_clients_should_be_able_to_exchange_messages_given_a_libp2p_network ... FAILED
```

Failure does not seem related to this PR.

### CODE REVIEW

File: signer/src/main.rs
Comment (AshtonStephens) [inline signer/src/main.rs:385]: I think we should make this configurable, because if we find that we have significant issues somehow when there's a Bitcoin block storm it would be good to be able to tweak this. 

I don't think it'll be a hard change to make; the config classes are pretty clear, and you can put it under the bitcoin config settings [here](https://github.com/stacks-network/sbtc/blob/main/signer/src/config/mod.rs#L101) named something like `bitcoin_processing_delay`, and you'll have access to it via the `ctx` variable.

File: signer/src/transaction_coordinator.rs
Comment (AshtonStephens) [inline signer/src/transaction_coordinator.rs:223]: If you add the delay to the config you can delete this change and access the configuration via the context within the `self` reference.

File: signer/src/transaction_coordinator.rs
Comment (AshtonStephens) [inline signer/src/transaction_coordinator.rs:1050]: I'm curious why you added this comment instead of just deleting `should_be_able_to_coordinate_signing_rounds`. Are you asking for confirmation that you can delete it?

**Review by AshtonStephens — CHANGES_REQUESTED** [review]
We should change the approach to have the the configuration file take in a `bitcoin_block_processing_delay_s` variable for two reasons:
1. We may later want to change it across signers for some reason we don't yet know
2. It will make the code cleaner because the areas that need this value already have access to the configurations, making it not necessary to add to any function contracts.

File: signer/src/transaction_coordinator.rs
Comment (cylewitruk) [inline signer/src/transaction_coordinator.rs:171]: At first I was going to write that it would be easier to have the delay here, but I actually like that putting it in `process_blocks()` _after_ the "are we the coordinator" check lets us skip the delay if we determine that we're not the coordinator.

File: signer/src/transaction_coordinator.rs
Comment (cylewitruk) [inline signer/src/transaction_coordinator.rs:154]: nit: you should be able to use `std::time::Duration`

File: signer/tests/integration/emily.rs
Comment (cylewitruk) [inline signer/tests/integration/emily.rs:406]: nit: should be able to use `std::time::Duration`

File: signer/tests/integration/transaction_coordinator.rs
Comment (cylewitruk) [inline signer/tests/integration/transaction_coordinator.rs:282]: nit: should be able to use `std::time::Duration` (already `use`'d in this file)

File: signer/src/main.rs
Comment (aldur) [reply signer/src/main.rs:385]: I like that. The reason I didn't do it already is that there seem to be a bunch of hard-coded values already, eg. here: https://github.com/stacks-network/sbtc/blob/c8c133d0d2e5cd8c572bb964c03ba0b25a1febef/signer/src/main.rs#L380.

Let me add it.

File: signer/src/transaction_coordinator.rs
Comment (aldur) [reply signer/src/transaction_coordinator.rs:1050]: Yep, gathering feedback but also, if in doubt, with tests more is better :)

File: signer/src/transaction_coordinator.rs
Comment (cylewitruk) [inline signer/src/transaction_coordinator.rs:258]: bug? I think this needs to be moved before the above block coordinating DKG since it depends on the Bitcoin chain tip (which is what we're worried that the other signers may not have updated yet hence the delay).

File: signer/src/main.rs
Comment (cylewitruk) [reply signer/src/main.rs:385]: Sounds good, but I don't think this is a value that we want signers freely messing around with, so we should be careful about how we handle it in the `default.toml` and config examples.

File: signer/src/config/default.toml
Comment (aldur) [inline signer/src/config/default.toml:154]: Setting this to 0, but we should have a non-zero default when shipping to prod.

File: signer/src/transaction_coordinator.rs
Comment (aldur) [inline signer/src/transaction_coordinator.rs:1061]: Not sure if there's a better way to do this.

Comment (cylewitruk) [reply signer/src/transaction_coordinator.rs:1061]: I think I'd make the delay config explicitly in seconds as a uint.

File: signer/src/transaction_coordinator.rs
Comment (AshtonStephens) [reply signer/src/transaction_coordinator.rs]: Resolved.

File: signer/src/transaction_coordinator.rs
Comment (AshtonStephens) [reply signer/src/transaction_coordinator.rs:1045]: Makes sense, resolved.

File: signer/src/main.rs
Comment (AshtonStephens) [reply signer/src/main.rs]: Resolved.

File: signer/src/config/mod.rs
Comment (cylewitruk) [inline signer/src/config/mod.rs:242]: I think i would make the config explicitly in seconds, with a validation ceiling of 300 or so (half the expected time of a bitcoin block); that'll keep things simpler.

**Review by AshtonStephens — APPROVED** [review]
Approving, but please wait for any necessary resolutions on @cylewitruk's feedback before merging :)

File: signer/src/config/mod.rs
Comment (cylewitruk) [reply signer/src/config/mod.rs:242]: Nevermind, I read too quickly, you're doing this already 😅

Comment (aldur) [reply signer/src/config/mod.rs:242]: Do you mean renaming the variable to `... _seconds`? Can add the ceiling, no problem.

File: signer/src/transaction_coordinator.rs
Comment (aldur) [reply signer/src/transaction_coordinator.rs]: Uh, I think you are right. Done.

File: signer/src/config/mod.rs
Comment (cylewitruk) [reply signer/src/config/mod.rs:242]: No I actually meant what you are doing, I missed your custom deserializer at a glance and briefly thought that this was expected to be provided as a textual duration 🙈

Yeah I think ensuring that we have a reasonable max that's within our expected bitcoin block time would be wise

Comment (aldur) [reply signer/src/config/mod.rs:242]: So I implanted it because I read this too late (4e85213). I am not 100% sure we want to have that cap though.

**Comment by aldur** [comment]
Thanks both for the comments! @cylewitruk PTAL

File: signer/src/config/serialization.rs
Comment (cylewitruk) [inline signer/src/config/serialization.rs:61]: So we're generally not considering "valid" invalid values as errors during deserialization, so I think you could get rid of this custom serializer and just use the `duration_seconds_deserializer`.

There's `impl Validatable for SignerConfig {` on line 245 where the validation should be plugged in. I think that might make this all a little cleaner :)

File: signer/src/config/default.toml
Comment (cylewitruk) [reply signer/src/config/default.toml:154]: Could you just create a new ticket for this in the milestone as a "don't forget" and link it in a `# TODO: xxx` here?

File: signer/src/transaction_coordinator.rs
Comment (cylewitruk) [reply signer/src/transaction_coordinator.rs:1061]: That's what chatgpt is for :rofl:

File: signer/src/config/default.toml
Comment (aldur) [reply signer/src/config/default.toml:162]: Done in https://github.com/stacks-network/sbtc/issues/715.

File: signer/src/config/mod.rs
Comment (cylewitruk) [reply signer/src/config/mod.rs:244]: Yeah I'm not sure on the number, but I can't think of a situation where we would ever want to exceed 15-30 secs, so this is probably okay as a failsafe just to make sure that the signer can never wait _longer_ than a typical bitcoin block (i.e. waiting into the next block).

File: signer/src/config/mod.rs
Comment (djordon) [inline signer/src/config/mod.rs:27]: It looks like the build is failing because this constant is missing docs (we have a lint to make sure we have documentation for everything. Would you mind addressing this in another PR? Also, would you mind renaming this to `MAX_BITCOIN_PROCESSING_DELAY_SECS` or something similar?

**Review by djordon — COMMENTED** [review]
In general, we do not merge a PR unless all tests and checks have successfully passed. In the future, please try to make sure CI is green before merging. This might be one of those things that is worth encoding at the repo level, just like resolving all discussions.

File: signer/src/config/mod.rs
Comment (aldur) [reply signer/src/config/mod.rs:27]: Ops, sorry! I have seen failing tests locally and attributed the CI failure to that. Let me fix that right away.

**Pull request merged**
